src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSolutionEventListener.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright(c) Microsoft.All Rights Reserved.Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     Responsible for listening to project loaded events and firing <see cref="IUnconfiguredProjectTasksService.PrioritizedProjectLoadedInHost"/> and 
+    ///     <see cref="IUnconfiguredProjectTasksService.ProjectLoadedInHost"/>.
+    /// </summary>
+    [Export(typeof(ILoadedInHostListener))]
+    internal class VsSolutionEventListener : OnceInitializedOnceDisposedAsync, IVsSolutionEvents, IVsPrioritizedSolutionEvents, ILoadedInHostListener
+    {
+        private readonly IVsService<IVsSolution> _solution;
+        private readonly IProjectThreadingService _threadingService;
+        private uint _cookie = VSConstants.VSCOOKIE_NIL;
+
+        [ImportingConstructor]
+        public VsSolutionEventListener(IVsService<SVsSolution, IVsSolution> solution, IProjectThreadingService threadingService)
+            : base(threadingService.JoinableTaskContext)
+        {
+            _solution = solution;
+            _threadingService = threadingService;
+        }
+
+        public Task StartListeningAsync()
+        {
+            return InitializeAsync(CancellationToken.None);
+        }
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            await _threadingService.SwitchToUIThread();
+
+            HResult result = _solution.Value.AdviseSolutionEvents(this, out _cookie);
+            if (result.Failed)
+                throw result.Exception;
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            await _threadingService.SwitchToUIThread();
```

Comment:
Wrap this in `if (initialized) { ... }`. I think you will currently throw if disposed before initialization has finished.

Reply:
Throw how?

Reply:
I don't think we'll throw or can't think of a circumstance that we'll throw. We'll unnecessarily switch to the UI thread, so I've added a check.

Reply:
Yeah I didn't notice the cookie is nil check. Thought you would unadvise without ever having advised in the first place.